Auto-enable basic logging for non-Quarkus tests

famod opened the pull request:
This is what I had in mind here: https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/LoggingSetupRecorder.2EhandleFailedStart.28.29

It's not pretty with all that `static` stuff, but the challenge is to kinda "publish" the info early enough that regular Quarkus logging setup is imminent (about to happen), _before_ the globally enabled `NonQuarkusTestLoggingEnabler` is executed (which is _before_ extensions registered via `@ExtendWith`).
There is also the challenge to include `QuarkusUnitTest` and others from `junit5-internal` because the way this is all set up activates the `NonQuarkusTestLoggingEnabler` also in Quarkus modules (e.g. in CI) that depend on `quarkus-junit5-internal` _and_ `quarkus-junit5`.

The state handling in that enabler relies on "Quarkus tests first" test class ordering, as is our `QuarkusTestProfileAwareClassOrderer`'s default behavior.

I tested this successfully in `extensions/vertx-http/deployment` (and added a non-Quarkus test).

The registration setup has two "side effects" (which I think are bearable):
- `junit.jupiter.extensions.autodetection.enabled=true` will also enable other extensions that are on the classpath (via ServiceLoader)
- if users exclude `junit5-properties` (#22172), this enabler will still be picked up in case the custom `junit-platform.properties` contains `junit.jupiter.extensions.autodetection.enabled=true`

Draft until:
- [ ] code is cleaned up and some Javadoc is added
- [ ] I have tested this in my current project
- [ ] I have verified nothing strange happens in continuous testing (it shouldn't)

Do we need something in the testing docs? Like a hint that we activate `junit.jupiter.extensions.autodetection.enabled`? Not sure.

Btw, I don't think this can be tested automatically in a reasonable way but I'm open to suggestions!

famod committed:
Auto-enable basic logging for non-Quarkus tests

famod commented on test-framework/junit5/src/main/java/io/quarkus/test/junit/NonQuarkusTestLoggingEnabler.java:
```diff
@@ -0,0 +1,32 @@
+        if (!initialized) {
+            if (!PropertyTestUtil.isLoggingSetupImminent()) {
+                System.err.println("!!! NonQuarkusTestLoggingEnabler");
+                IntegrationTestUtil.activateLogging();
```
I would rather call `LoggingSetupRecorder` directly, but `handleFailedStart()` sounds weird.

Given that it's used in various non-failure cases, WDYT about renaming `handleFailedStart()` to `initializeBasicLogging()`? (or deprecate for removal and add the new method)

geoand replied:
Yeah, that makes sense

famod replied:
Ok, do you think it's used outside of this repo? If yes we should play it safe and deprecate.

geoand replied:
I doubt it's used, but let's play it safe just in case

geoand commented:
Cool!

famod commented:
This looks interesting:
```
2022-01-12T00:09:41.2592984Z [INFO] Running io.quarkus.restclient.config.RestClientConfigTest
2022-01-12T00:09:41.4205046Z !!! NonQuarkusTestLoggingEnabler
2022-01-12T00:09:41.5232793Z [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.257 s <<< FAILURE! - in io.quarkus.restclient.config.RestClientConfigTest
2022-01-12T00:09:41.5237369Z [ERROR] io.quarkus.restclient.config.RestClientConfigTest.testLoadRestClientConfig  Time elapsed: 0.092 s  <<< ERROR!
2022-01-12T00:09:41.5240044Z java.lang.IllegalStateException: SRCFG00017: Configuration already registered for the given class loader
2022-01-12T00:09:41.5242627Z 	at io.smallrye.config.SmallRyeConfigProviderResolver.registerConfig(SmallRyeConfigProviderResolver.java:118)
2022-01-12T00:09:41.5245960Z 	at io.quarkus.restclient.config.RestClientConfigTest.setupMPConfig(RestClientConfigTest.java:70)
2022-01-12T00:09:41.5249471Z 	at io.quarkus.restclient.config.RestClientConfigTest.testLoadRestClientConfig(RestClientConfigTest.java:25)
```

famod commented:
@geoand @radcortez so that test fails here: https://github.com/quarkusio/quarkus/blob/ad4ab21c16b10cfce0c6ccce36bc68bf27d1ef3c/extensions/resteasy-classic/rest-client-config/runtime/src/test/java/io/quarkus/restclient/config/RestClientConfigTest.java#L70
because apparently somewhere [here](https://github.com/quarkusio/quarkus/blob/2b4496718ec46836e2775f669c7c8c5f970ccaa1/core/runtime/src/main/java/io/quarkus/runtime/logging/LoggingSetupRecorder.java#L71-L80) the `LoggingSetupRecorder` seems to be registering a config already.
I could just add a flag to that test so that the enabler doesn't run, but I'm under the impression that the recorder is missing a release of the configs it created?

geoand commented:
Also cc @michalszynkiewicz for ^

famod commented:
I suppose we could also release any existing config in the test before it tries to register its own.
Still, if that band aid recorder method is leaking something, we should probably fix that instead.

radcortez commented:
> @geoand @radcortez so that test fails here:
> 
> https://github.com/quarkusio/quarkus/blob/ad4ab21c16b10cfce0c6ccce36bc68bf27d1ef3c/extensions/resteasy-classic/rest-client-config/runtime/src/test/java/io/quarkus/restclient/config/RestClientConfigTest.java#L70
> 

Yes, this test should release the `Config` from the Classloader in the end, or in the alternative, build one that is not associated with the current Classloader.

famod commented:
I think I will try to do "both":
- try to release any config that `LoggingSetupRecorder` is leaving behind
- release the config the test is creating (after it's done)

famod commented:
So I've fixed that test by releasing the config that the recorder has "left behind". I'm doing this only in the enabler, not the recorder, because I wanted to limit the side effects and apparently, there haven't been any issues with existing calls to the recorder.

I also added a config release to that test (so that cleans up after itself) with the same pattern that's used in some core config tests, e.g. `UUIDPropertyTest`.
@radcortez what I don't get (but I still copied it) is this: https://github.com/quarkusio/quarkus/blob/7dc9eeeace25f4f39cd031b0f4e167b91e213c5c/core/runtime/src/test/java/io/quarkus/runtime/configuration/UUIDPropertyTest.java#L30-L34
While testing, I realized that you will _always_ get a config instance from `ConfigProviderResolver.getConfig()`, no matter whether the test registered one or not. So I'm not getting when that ISE can actually occur? 🤔

radcortez commented:
> While testing, I realized that you will _always_ get a config instance from `ConfigProviderResolver.getConfig()`, no matter whether the test registered one or not. So I'm not getting when that ISE can actually occur? 🤔

When you call `ConfigProviderResolver.getConfig()`, this will create a new `Config` associated with the current classloader. So if you call `ConfigProviderResolver.instance().registerConfig` you get the ISE if a `Config` is already associated. This may happen in the `RestClientConfigTest`, if a previously executed test left the `Config` in the cache:

https://github.com/quarkusio/quarkus/blob/9bba11de3af072eaf60836dd3e58cc929552228b/extensions/resteasy-classic/rest-client-config/runtime/src/test/java/io/quarkus/restclient/config/RestClientConfigTest.java#L64-L71

Actually, the safest way for testing is to just create the `Config` using the builder, because it will not be cached, but usually, people just call `ConfigProvider.getConfig()`.

famod commented:
@radcortez 
> So if you call `ConfigProviderResolver.instance().registerConfig` you get the ISE if a `Config` is already associated. 

The _register_ case is clear to me, but isn't the bit I posted about an ISE when _releasing_?

kdubb commented:
@famod Do you think this would be a workaround/fix for #22844? It's a pretty serious issue with Gradle test logging. It seems this PR would allow skipping the setup of the `LogManager` via system property.

radcortez commented:
> @radcortez
> 
> > So if you call `ConfigProviderResolver.instance().registerConfig` you get the ISE if a `Config` is already associated.
> 
> The _register_ case is clear to me, but isn't the bit I posted about an ISE when _releasing_?

Ah yes, for release it doesn't make sense.

famod commented:
@radcortez ok, then I'll remove the catch and create another PR to remove it from the core tests (or do you want me to do it here instead?)

@kdubb
> It seems this PR would allow skipping the setup of the `LogManager` via system property.

No, this PR is about setting up basic logging for a non-Quarkus test if no Quarkus test has been run before.
The system property is not supposed to be set by anyone... but as I've been thinking about a kill switch for this featue, I might actually reconsider 🤔 - but anyway, I don't think this is of any use for you.

kdubb commented:
@famod The "system property" I was referring to was the one that is suggested to be set to install the LogManager during testing in Gradle (e.g. setting `java.util.logging.manager` to `org.jboss.logmanager.LogManager`).

radcortez commented:
> @radcortez ok, then I'll remove the catch and create another PR to remove it from the core tests (or do you want me to do it here instead?)

In another PR is fine.

famod commented:
@geoand so I've hit a problem: After an e.g. `QuarkusUnitTest`, `Timing` will remove all log handlers and will effectively disable the `QuarkusDelayedHandler`:
https://github.com/quarkusio/quarkus/blob/0ffe2276e7b3b80c352cca4a7d087caebc1ffbe3/independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/Timing.java#L126
When there is a non-Quarkus test after this Quarkus test, the enabler will not run (sp far so expected).

The problem here is that any logging issued by that non-Quarkus test will be _delayed_ until the next Quarkus test sets up proper logging again, reactivating the `QuarkusDelayedHandler`.
It's very confusing because at the time the message is actually printed the respective test is long done.

The only way out of this AFAICS is to set up basic logging for *each* non-Quarkus test class (if `QuarkusDelayedHandler` is not active, which will be usually the case here).
It seems I'm back to checking fields and annotations, yikes! (unless you have another idea)

famod commented:
...or we _always_ set up basic logging in that enabler (unless the handler is already active).
This way we might even be able to use logging in places where we today have to use sysout instead.

geoand commented:
I kinda feel that this is getting too complicated...

famod commented:
The implementation would actually be easier, AFAICS. But there will be small runtime penalty.

Let me create another draft PR...

famod commented:
I'd very much prefer #22920; pretty simple and gets the job done (without that delay problem).

famod commented:
Superseded by #22920.
